adapter/oci/app_client.go

```diff
@@ -0,0 +1,208 @@
+	}
+
+	for _,s := range subnetsArray {
+		// Typecast to string
+		subnetString, secondSuccess := s.(string)
+		if !secondSuccess {
+			return nil, errors.New("Invalid subnets annotation")
+		}
+		subnets = append(subnets, subnetString)
+	}
+
+	return subnets, nil
+}
+
+func (a AppClient) GetApp(appName string) (*adapter.App, error) {
+	compartmentId := viper.GetString("oracle.compartment-id")
+	req := functions.ListApplicationsRequest{CompartmentId: &compartmentId, DisplayName: &appName}
+	resp, err := a.client.ListApplications(context.Background(), req)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.Items) > 0 {
+		getreq := functions.GetApplicationRequest{ApplicationId: resp.Items[0].Id}
+		getres,geterr := a.client.GetApplication(context.Background(), getreq)
+
+		if geterr != nil {
+			return nil, geterr
+		}
+
+		adapterApp := convertOCIAppToAdapterApp(&getres.Application)
+		return adapterApp, nil
+	} else {
+		return nil, adapter.AppNameNotFoundError{Name: appName}
+	}
+}
+
+func (a AppClient) UpdateApp(app *adapter.App) (*adapter.App, error) {
+	//merge config
+	for k, v := range app.Config {
+		if v == "" {
+			delete(app.Config, k)
+		} else {
+			app.Config[k] = v
+		}
+	}
+
+	body := functions.UpdateApplicationDetails{
+		Config: app.Config,
+	}
+
+	req := functions.UpdateApplicationRequest{UpdateApplicationDetails: body, ApplicationId: &app.ID}
+	res, err := a.client.UpdateApplication(context.Background(), req)
+
+	if err != nil {
+		return nil, err
+	}
+
+	adapterApp := convertOCIAppToAdapterApp(&res.Application)
+	return adapterApp, nil
+}
+
+func (a AppClient) DeleteApp(appID string) error {
+	req := functions.DeleteApplicationRequest{ApplicationId: &appID,}
+	_, err := a.client.DeleteApplication(context.Background(), req)
+	return err
+}
+
+func (a AppClient) ListApp(limit int64) ([]*adapter.App, error) {
+	compartmentId := viper.GetString("oracle.compartment-id")
+	var resApps []*adapter.App
+	req := functions.ListApplicationsRequest{CompartmentId: &compartmentId,}
+
+	for {
+		resp, err := a.client.ListApplications(context.Background(), req)
+		if err != nil {
+			return nil, err
+		}
+
+		adapterApps := convertOCIAppsToAdapterApps(&resp.Items)
+		resApps = append(resApps, adapterApps...)
+		howManyMore := limit - int64(len(resApps)+len(resp.Items))
+
+		if howManyMore <= 0 || resp.OpcNextPage == nil {
+			break
+		}
+
+		req.Page = resp.OpcNextPage
+	}
+
+	if len(resApps) == 0 {
+		fmt.Fprint(os.Stderr, "No apps found\n")
+		return nil, nil
+	}
+
+	return resApps, nil
+}
+
+func convertOCIAppsToAdapterApps(ociApps *[]functions.ApplicationSummary) []*adapter.App {
+	var resApps []*adapter.App
+	for _, ociApp := range *ociApps {
+		app := convertOCIAppSummaryToAdapterApp(&ociApp)
+		resApps = append(resApps, app)
+	}
+
+	return resApps
+}
+
+func convertOCIAppSummaryToAdapterApp(ociApp *functions.ApplicationSummary) *adapter.App {
+	createdAt, _ := strfmt.ParseDateTime(ociApp.TimeCreated.Format(time.RFC3339Nano))
```

Comment:
Should we be checking and handling `err` here?

Reply:
Nice catch! Fixed it. Thanks.